File: repo/backend/app/schemas/message_center.py
from datetime import datetime

from pydantic import BaseModel, Field


class MessageTemplateCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=180)
    category: str = Field(min_length=1, max_length=80)
    channel: str = Field(default="in_app", min_length=1, max_length=20)
    body_template: str = Field(min_length=1, max_length=20000)
    is_active: bool = True


class MessageTemplateUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=180)
    category: str | None = Field(default=None, min_length=1, max_length=80)
    channel: str | None = Field(default=None, min_length=1, max_length=20)
    body_template: str | None = Field(default=None, min_length=1, max_length=20000)
    is_active: bool | None = None


class MessageTemplateOut(BaseModel):
    id: str
    project_id: str
    name: str
    category: str
    channel: str
    body_template: str
    variables: list[str]
    is_active: bool
    created_at: datetime
    updated_at: datetime


class MessagePreviewRequest(BaseModel):
    template_id: str
    itinerary_id: str | None = None
    variables: dict[str, str] = Field(default_factory=dict)


class MessagePreviewOut(BaseModel):
    template_id: str
    rendered_body: str
    missing_variables: list[str]
    variables_used: dict[str, str]


class MessageSendRequest(BaseModel):
    template_id: str
    recipient_user_id: str = Field(min_length=1, max_length=180)
    itinerary_id: str | None = None
    variables: dict[str, str] = Field(default_factory=dict)


class MessageDeliveryAttemptOut(BaseModel):
    id: str
    connector_key: str
    attempt_status: str
    provider_message_id: str | None
    detail: str | None
    response_payload: dict | None
    attempted_at: datetime


class MessageDispatchOut(BaseModel):
    id: str
    project_id: str
    itinerary_id: str | None
    template_id: str | None
    template_name: str
    template_category: str
    channel: str
    recipient_user_id: str
    recipient_display_name: str | None
    rendered_body: str
    send_status: str
    variables_payload: dict[str, str]
    created_by_user_id: str
    created_at: datetime
    attempts: list[MessageDeliveryAttemptOut]

File: repo/frontend/src/api/message-center.ts
import { apiGet, apiPatch, apiPost } from "./client";

export type MessageTemplate = {
  id: string;
  project_id: string;
  name: string;
  category: string;
  channel: string;
  body_template: string;
  variables: string[];
  is_active: boolean;
  created_at: string;
  updated_at: string;
};

export type MessagePreview = {
  template_id: string;
  rendered_body: string;
  missing_variables: string[];
  variables_used: Record<string, string>;
};

export type MessageDeliveryAttempt = {
  id: string;
  connector_key: string;
  attempt_status: string;
  provider_message_id: string | null;
  detail: string | null;
  response_payload: Record<string, unknown> | null;
  attempted_at: string;
};

export type MessageDispatch = {
  id: string;
  project_id: string;
  itinerary_id: string | null;
  template_id: string | null;
  template_name: string;
  template_category: string;
  channel: string;
  recipient_user_id: string;
  recipient_display_name: string | null;
  rendered_body: string;
  send_status: string;
  variables_payload: Record<string, string>;
  created_by_user_id: string;
  created_at: string;
  attempts: MessageDeliveryAttempt[];
};

export const messageCenterApi = {
  listTemplates: (projectId: string) => apiGet<MessageTemplate[]>(`/api/projects/${projectId}/message-center/templates`),

  createTemplate: (
    projectId: string,
    payload: {
      name: string;
      category: string;
      channel: string;
      body_template: string;
      is_active: boolean;
    }
  ) => apiPost<MessageTemplate>(`/api/projects/${projectId}/message-center/templates`, payload),

  updateTemplate: (
    projectId: string,
    templateId: string,
    payload: {
      name?: string;
      category?: string;
      channel?: string;
      body_template?: string;
      is_active?: boolean;
    }
  ) => apiPatch<MessageTemplate>(`/api/projects/${projectId}/message-center/templates/${templateId}`, payload),

  preview: (
    projectId: string,
    payload: {
      template_id: string;
      itinerary_id?: string | null;
      variables: Record<string, string>;
    }
  ) => apiPost<MessagePreview>(`/api/projects/${projectId}/message-center/preview`, payload),

  send: (
    projectId: string,
    payload: {
      template_id: string;
      recipient_user_id: string;
      itinerary_id?: string | null;
      variables: Record<string, string>;
    }
  ) => apiPost<MessageDispatch>(`/api/projects/${projectId}/message-center/send`, payload),

  listTimeline: (projectId: string, limit = 50) =>
    apiGet<MessageDispatch[]>(`/api/projects/${projectId}/message-center/timeline?limit=${encodeURIComponent(String(limit))}`)
};
